src/client/driver/command-executors/execute-action-command.js

```diff
@@ -2,76 +2,107 @@ import hammerhead from '../deps/hammerhead';
-var DblClickAutomation = testCafeRunner.get('./automation/playback/dblclick');
-var HoverAutomation    = testCafeRunner.get('./automation/playback/hover');
-var ProgressPanel      = testCafeUI.ProgressPanel;
+var Promise                 = hammerhead.Promise;
+var XhrBarrier              = testCafeCore.XhrBarrier;
+var pageUnloadBarrier       = testCafeCore.pageUnloadBarrier;
+var positionUtils           = testCafeCore.positionUtils;
+var waitFor                 = testCafeCore.waitFor;
+var ClickAutomation         = testCafeRunner.get('./automation/playback/click');
+var RClickAutomation        = testCafeRunner.get('./automation/playback/rclick');
+var DblClickAutomation      = testCafeRunner.get('./automation/playback/dblclick');
+var DragToOffsetAutomation  = testCafeRunner.get('./automation/playback/drag/to-offset');
+var DragToElementAutomation = testCafeRunner.get('./automation/playback/drag/to-element');
+var HoverAutomation         = testCafeRunner.get('./automation/playback/hover');
+var ProgressPanel           = testCafeUI.ProgressPanel;
 
 
 const PROGRESS_PANEL_TEXT   = 'Waiting for the target element of the next action to appear';
 const CHECK_ELEMENT_DELAY   = 200;
 const CHECK_ELEMENT_TIMEOUT = 10000;
 
 
-function ensureElementExists (selector) {
+function ensureElementExists (selector, ErrorCtor) {
     return waitFor(selector, CHECK_ELEMENT_DELAY, CHECK_ELEMENT_TIMEOUT)
         .catch(() => {
-            throw new ActionElementNotFoundError();
+            throw new ErrorCtor();
         });
 }
 
-function ensureElementVisible (element, timeout) {
+function ensureElementVisible (element, timeout, ErrorCtor) {
     return waitFor(() => positionUtils.isElementVisible(element) ? element : null, CHECK_ELEMENT_DELAY, timeout)
         .catch(() => {
-            throw new ActionElementIsInvisibleError();
+            throw new ErrorCtor();
         });
 }
 
-function ensureElement (selector) {
-    var startTime     = new Date();
-    var progressPanel = new ProgressPanel();
-
-    progressPanel.show(PROGRESS_PANEL_TEXT, CHECK_ELEMENT_TIMEOUT);
+function ensureElement (selector, NotFoundErrorCtor, IsInvisibleErrorCtor) {
+    var startTime = new Date();
 
-    return ensureElementExists(selector, CHECK_ELEMENT_TIMEOUT)
+    return ensureElementExists(selector, NotFoundErrorCtor)
         .then(element => {
             var checkVisibilityTimeout = CHECK_ELEMENT_TIMEOUT - (new Date() - startTime);
 
-            return ensureElementVisible(element, checkVisibilityTimeout);
-        })
-        .then(element => {
-            progressPanel.close(true);
-            return element;
-        })
-        .catch(err => {
-            progressPanel.close(false);
-            throw err;
+            return ensureElementVisible(element, checkVisibilityTimeout, IsInvisibleErrorCtor);
+        });
+}
+
+function ensureSelectors (command) {
+    if (command.type === COMMAND_TYPE.dragToElement) {
```

Comment:
Avoid having inconsistent function results based on input value. Rather have two functions, or call `ensureElement` for each argument and pass error Ctor (use standard error Ctor as default value)